### PR TITLE
Usage of FontFace rather than paths

### LINKED ISSUE #1127: Improvements to Glyph Rendering in SVGContext / CanvasContext
This thread has evolved into discussing ways to improve Glyph rendering. I'll keep this issue open for now, even though there is no bug to fix. If we prefer to move it to the GitHub discussions tab, I can close off this issue.

---

Original title:
## Should OutlineCode enum (in glyph.ts) use ints other than 0, 1, 2, 3?

This isn't a _bug_, but I was reading glyph.ts and saw that `GlyphOutline.parse()` returns a `number[]` where the **_position_** of the number changes its meaning.

For example, Bach Demo produces an array like:
`[0,0,60,3,140,180,0,135,62,180,3,425,60,268,180,425,62,3,285,180,425,134,367,180,3,0,60,127,180,0,63]`

The first 0 is actually a "move to" instruction, and the second 0 is an x coordinate with value 0, and then comes a y coordinate with value 60. The 3 after that starts a bezier curve....

Then, `Glyph.renderOutline()` and `Glyph.getOutlineBoundingBox()` loop over these outline arrays to draw the glyphs to the context.

```
    while (!go.done()) {
      switch (go.next()) {
        case OutlineCode.MOVE: // => 0
          ctx.moveTo(go.nextX(), go.nextY());
          break;
        case OutlineCode.LINE: // => 1
          ctx.lineTo(go.nextX(), go.nextY());
          break;
          ...
```

So the switch statement is looking for numbers 0, 1, 2, 3.

Anyways, like I said, this isn't a bug because the three methods all follow the same convention (a number 0 implies an x, y and a number 3 is followed by 6 ints specifying the points of a cubic bezier curve).

I know it was part of @tommadams massive improvements on glyph performance & correctness in https://github.com/0xfe/vexflow/pull/1103 and  https://github.com/0xfe/vexflow/pull/1077.

Obviously we don't want to go back to the old / slow way. 

But do we have any ideas on if / how we could improve this, or make it potentially less brittle / future proof?

I can't think of anything off the top of my head at the moment, other than starting the enum at a really big number (that is still well under `Number.MAX_SAFE_INTEGER`), like:

```
export const enum OutlineCode {
  MOVE = 0xfffffffffff0,
  LINE = 0xfffffffffff1,
  QUADRATIC = 0xfffffffffff2,
  BEZIER = 0xfffffffffff3,
}
```

Anyways, can we improve this in any way? If not, I'll close as "not a bug."

### LINKED ISSUE #1459: Measuring/estimating text in VexFlow quickly and accurately
I started this with a specific issue in mind, but this quickly became a larger discussion in some PRs, so we are moving the discussion here.

### Introduction
In many cases we need to know metrics of text font files to place and render them with music glyphs.  Drawing and measuring the text has some problems: 1) during the format part of the pipeline, the rendering context isn't available, and 2) rendering anything on the VF renderer canvas/svg can cause a reflow which will greatly slow down the rendering of large scores (especially in SVG).

### Summary of the Proposals
There are a couple of solutions to this that have been introduced, each has its own drawbacks:
1. TextFormatter is a class that uses font metrics derived directly from the font, using the same tools we use to derive music font metrics.  It was introduced with the ChordSymbol module.  Drawbacks: 1) you need to have the metrics for the font _a_ _priori_, if the metrics aren't available the accuracy won't be as good.  And 2) representing information like kerning pairs could result in some large, static files.
2. @rvilarl had the idea to create a canvas context specifically for the purpose of rendering the bounding box.  Since this isn't the canvas where the score is rendered, it shouldn't cause a reflow and will does not require the text metrics to be derived in advance.  Drawbacks: 1) lazy-loaded fonts might not be available when rendering (this is the issue that initially started this thread)  2) we need to investigate if there are cases where canvas context would not be available and text estimation needs to be done.  Right now there is only one DOM element required to use VexFlow, when the renderer is created.

### PR DESCRIPTION
**Pull request by rvilarl**
fixes #1127
fixes #1459

Personally I see two main benefits:
- the library is smaller.
- the produced SVG are also smaller.

```
-rw-r--r--. 1 rodrigo rodrigo 1911241 nov 27 22:06 reference/cjs/vexflow-debug.js
-rw-r--r--. 1 rodrigo rodrigo 1380174 nov 27 22:15 build/cjs/vexflow-debug.js
```
StaveNote with two noteheads with one accidental each.
```svg
<g class="vf-stavenote" id="vf-auto1300">
 <g class="vf-notehead" id="vf-auto1348" pointer-events="bounding-box">
  <text stroke="none" x="258.331" y="120"></text>
  <text stroke="none" x="232.347" y="120"></text>
 </g>
 <g class="vf-notehead" id="vf-auto1349" pointer-events="bounding-box">
  <text stroke="none" x="258.331" y="110"></text>
  <text stroke="none" x="245.339" y="110"></text>
 </g>
</g>
```

### CODE REVIEW

**Comment by ronyeh** [comment]
As this is a big PR, it'll take time to properly review. Thanks for looking into this though!

Can you write up some extra info in this thread to help us with this review? What are the main tradeoffs?

This handles all of our font needs, whether we are using it to render chord symbols or music engraving symbols??

**Comment by rvilarl** [comment]
> As this is a big PR, it'll take time to properly review. Thanks for looking into this though!
> 
> Can you write up some extra info in this thread to help us with this review? What are the main tradeoffs?

The changes are simple and the **review will be much quicker than what you anticipate** :)

- Concept: `xxx_glyphs.ts` files containing no path ('o' member) contains now a 'unicode' value. These unicodes will be printed with `FillText`.
- `Font` has two new methods to measure text: `textMetrics` and `measureText`. Those replace `TextFormatter`. This is related to most of the changes. This method is quick enough and a text cache is no longer required.
- Images generation with `jsdom`, I only managed to get the fonts loaded by node using `RegisterFont`.
 
> This handles all of our font needs, whether we are using it to render chord symbols or music engraving symbols??

Yes both, and also, to write text. The SMuFL OTF files include  characters for many languages (hindi,...)

**Comment by mscuthbert** [comment]
I don't ever manipulate the points of the path after its drawn, so that won't be a breaking change to me, but I do tend to change the fill-color (and opacity, etc.) of the SVG.  Is that done with the same interface in this PR?  It looks like an incredible  space savings (especially since I can then load one copy of Bravura [or Gootville] for the whole site instead of two, but I wonder if it's backwards compatible enough to go into v4 or if it's more of a v5 thing?

**Comment by rvilarl** [comment]
Draft until #1476 is resolved, I want to integrate this PR in this one.

File: src/annotation.ts [hunk 1]
```diff
@@ -84,15 +83,15 @@ export class Annotation extends Modifier {
     let maxRightGlyphWidth = 0;
     for (let i = 0; i < annotations.length; ++i) {
       const annotation = annotations[i];
-      const textFormatter = TextFormatter.create(annotation.textFont);
+      const measure = Font.measureText(annotation.text, annotation.textFont!);
```
Comment (mscuthbert) [inline src/annotation.ts:86]: perhaps refactor from `measure` to `measurement`?  `measure` has a very specific implication in music applications.

**Pull request closed without merging**